### PR TITLE
Allow PintoHelixResourceManager subclasses to be used in the controller starter by providing an overridable PinotHelixResouceManager object creator function

### PR DESCRIPTION
**Pull request by 9aman**
### Current Scenario
In `BaseControllerStarter`, the `PinotHelixResourceManager _helixResourceManager` initialization is hard coded to the PinotHelixResourceManager object.

This reduces the flexibility of using **subclasses** of PinotHelixResourceManager in any subclass of the `BaseControllerStarter`. 

### Changes in the PR
Introduce an overridable function for initialization of PinotHelixResourceManager **_helixResourceManager**. Thus, any **subclass** of PinotHelixResourceManager can be used for initialization in the subclass of `BaseControllerStarter`.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 1]
```diff
@@ -215,15 +215,15 @@ private enum LineageUpdateType {
...
+  protected final String _dataDir;
   private final boolean _isSingleTenantCluster;
   private final boolean _enableBatchMessageMode;
   private final int _deletedSegmentsRetentionInDays;
   private final boolean _enableTieredSegmentAssignment;
 
   private HelixManager _helixZkManager;
   private HelixAdmin _helixAdmin;
-  private ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  protected ZkHelixPropertyStore<ZNRecord> _propertyStore;
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:226]: there is getPropertyStore(), so no need to make this protected.

you added getter for _dataDir, so no need to make it protected either, in fact, _dataDir is from _config which is passed in to the constructor, so you may get it from _config.

File: pinot-controller/src/main/java/org/apache/pinot/controller/BaseControllerStarter.java [hunk 2]
```diff
@@ -297,6 +297,18 @@ private void setupHelixClusterConstraints() {
...
+   * <p>
+   * This method can be overridden by subclasses to instantiate the object
+   * with subclasses of PinotHelixResourceManager.
+   * By default, it returns a new PinotHelixResourceManager using the current configuration.
+   *
+   * @return A new instance of PinotHelixResourceManager.
+   */
+  protected PinotHelixResourceManager createPinotHelixResourceManager() {
+    return new PinotHelixResourceManager(_config);
+  }
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/BaseControllerStarter.java:311]: (nit) add an empty line below

File: pinot-controller/src/main/java/org/apache/pinot/controller/BaseControllerStarter.java [hunk 3]
```diff
@@ -297,6 +297,18 @@ private void setupHelixClusterConstraints() {
...
+  /**
+   * Creates an instance of PinotHelixResourceManager.
+   * <p>
+   * This method can be overridden by subclasses to instantiate the object
+   * with subclasses of PinotHelixResourceManager.
+   * By default, it returns a new PinotHelixResourceManager using the current configuration.
+   *
+   * @return A new instance of PinotHelixResourceManager.
+   */
+  protected PinotHelixResourceManager createPinotHelixResourceManager() {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/BaseControllerStarter.java:309]: (minor) To be consistent with getter
```suggestion
  protected PinotHelixResourceManager createHelixResourceManager() {
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java (hunk 1 above)
Comment (9aman) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:226]: Will use the getter.

**Review by klsince — APPROVED** [review]

**Pull request merged**